Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,129 @@
+* Selectors for service, probe, ScrapeConfig
+* Replica
+* Shard
+* Storage
+
+We will add a new `mode` field that accepts either `StatefulSet` or `DaemonSet`, with `StatefulSet` being the default. If the DaemonSet mode is activated (`mode: DaemonSet`), all the unrelated fields listed above will not be accepted. In the MVP, we will simply fail the reconciliation if any of those fields are set. We will prevent users from directly switching from a live StatefulSet setup to DaemonSet, because that might break their workload if they forget to unset the unsupported fields. Following up, we will leverage validation rules with [Kubernetes' Common Expression Language (CEL)](https://kubernetes.io/docs/reference/using-api/cel/). Only then, we will allow switching from a live StatefulSet setup to DaemonSet. We already have an issue for CEL [here](https://github.com/prometheus-operator/prometheus-operator/issues/5079).
+
+### 6.2. Node detecting:
+
+As pointed out in [Danny from GMP’s talk](https://www.youtube.com/watch?v=yk2aaAyxgKw), to make Prometheus Agent DaemonSet know which node it’s on, we can use [Kubernetes’ downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/). In `config-reloader` container, we can mount the node name as an environment variable like this:
+```
+containers:
+- name: config-reloader
+  env:
+  - name: NODE_NAME
+    valueFrom:
+      fieldRef:
+        apiVersion: v1
+        fieldPath: spec.nodeName
+```
+
+### 6.3. Targets filtering for pods (PodMonitor support):
+
+To filter pod targets, Danny’s talk has pointed out the best option is to use field selector:
+```
+kubernetes_sd_configs:
+- role: pod
+  selectors:
+  - role: pod
+    field: spec.nodeName=$NODE_NAME
+```
+We'll go with this option, because it filters targets right at discovery time, and also because Kubernetes API server watch cache indexes pods by node name (as we can see in [Kubernetes codebase](https://github.com/kubernetes/kubernetes/blob/v1.30.0-rc.0/pkg/registry/core/pod/storage/storage.go#L91)).
+
+We've also considered using relabel config that filters pods by `__meta_kubernetes_pod_node_name` label. However, we didn't choose to go with this option because it filters pods only after discovering all the pods from PodMonitor, which increases load on Kubernetes API server.
+
+
+## 7. Action Plan
+
+For the implementation, we’ll do what we detailed in the How section. The common logics between StatefulSet and DaemonSet modes will be extracted into one place. We will have a separate `daemonset.go` for the separate logic of the DaemonSet mode.
+
+For the test, we will have unit tests covering new logic, and integration tests covering the basic user cases, which are:
+* Prometheus Agent DaemonSet cannot be deployed when a similar Prometheus Agent StatefulSet resource is there.
```

Comment:
Hmmm, my understanding was that we wanted to allow statefulset and deamonset to run together, as long as they are deployed as different resources.

Have we changed our mind recently?

Reply:
Yes we allow different resources of StatefulSet and DaemonSet together. I stated here that we don't allow _**similar**_ StatefulSet and DaemonSet resources. Similar means scraping same set of targets. Is that correct understanding? If yes, I'll clarify it in the proposal.

Reply:
Could you explain why we won't allow them to scrape the same targets?

Reply:
As we discussed, in MVP, we won't allow directly switching from a live StatefulSet to DaemonSet. AFAIU we would check if the DaemonSet resource has the same set of targets with a live StatefulSet. If yes, we'll fail the DaemonSet. Is that correct understanding?

And as @simonpasquier described in https://github.com/prometheus-operator/prometheus-operator/pull/6600#discussion_r1617338651, seems like we won't allow StatefulSet and DaemonSet to scrape same set of targets even after MVP?

Reply:
Yeah, agreed that initially we won't allow switching from DaemonSet to StatefulSet (and vice-versa), but my understanding from the current wording is that we won't allow separate agents (one sts and another daemonset) to scrape the same targets.

I don't remember us agreeing on the latter. If someone wants sts and daemonset scraping the same targets, as long as they are deployed independently, why would we block it?

Reply:
Ok agree that we never discussed about blocking this. What @simonpasquier described in https://github.com/prometheus-operator/prometheus-operator/pull/6600#discussion_r1617338651 is just his mixed deployment case.

So back to my question in https://github.com/prometheus-operator/prometheus-operator/pull/6600#discussion_r1615270423: In case user is using StatefulSet, and they want to deploy an additional DaemonSet scraping same set of targets, they would first have to delete the existing StatefulSet, then deploy DaemonSet, then deploy StatefulSet again. Is that correct? 🤔

Reply:
We've discussed offline and clarified that we don't want to limit users. It will stay the same as today: it's ok to have 2 Prometheus resources scraping the same set of targets.